File: internal/cloud/portal/config_test.go
package portal

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"testing"

	"github.com/JoseAFlores777/ccp/internal/core"
)

// maquinaFalsa monta lo justo para que core.SnapshotSources devuelva una
// configuración con todas sus capas: ccp, lo global, dos perfiles, las
// ventanas de Desktop y un repo con regla.
func maquinaFalsa(t *testing.T) (home, src string) {
	t.Helper()
	home = t.TempDir()
	src = filepath.Join(t.TempDir(), ".claude")
	desktop := t.TempDir()
	t.Setenv("CCP_CLAUDE_SRC", src)
	t.Setenv("CCP_DESKTOP_DEFAULT_DATA_DIR", desktop)
	t.Setenv("HOME", t.TempDir())
	write := func(p, s string, mode os.FileMode) {
		t.Helper()
		if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
			t.Fatal(err)
		}
		if err := os.WriteFile(p, []byte(s), mode); err != nil {
			t.Fatal(err)
		}
	}
	write(filepath.Join(src, "settings.json"), `{"model":"opus"}`, 0o644)
	write(filepath.Join(src, "CLAUDE.md"), "# global\n", 0o644)
	write(filepath.Join(src, "keybindings.json"), `{}`, 0o644)
	write(filepath.Join(src, "agents", "revisor.md"), "# revisor\n", 0o644)
	write(filepath.Join(src, "commands", "x.md"), "# x\n", 0o644)
	write(filepath.Join(src, "skills", "s", "SKILL.md"), "# s\n", 0o644)
	write(filepath.Join(src, "output-styles", "e.md"), "# e\n", 0o644)
	write(filepath.Join(src, "hooks", "h.sh"), "#!/bin/sh\n", 0o755)
	write(filepath.Join(src, "plugins", "installed_plugins.json"), `{"plugins":{}}`, 0o644)
	write(filepath.Join(src, "plugins", "known_marketplaces.json"), `{}`, 0o644)
	write(src+".json", `{"mcpServers":{"a":{"command":"node"}}}`, 0o600)
	write(filepath.Join(desktop, "claude_desktop_config.json"), `{"mcpServers":{}}`, 0o600)

	if err := core.ProfileAddOfficial(home, "work"); err != nil {
		t.Fatal(err)
	}
	if err := core.CfgInitOverlay(home, "work"); err != nil {
		t.Fatal(err)
	}
	if err := core.ProfileAddDeepseek(home, "deep", core.BuiltinDefaults()); err != nil {
		t.Fatal(err)
	}
	if err := core.ProfileSetKey(home, "deep", "sk-1"); err != nil {
		t.Fatal(err)
	}
	repo := t.TempDir()
	write(filepath.Join(repo, ".git", "config"),
		"[core]\n\tbare = false\n[remote \"origin\"]\n\turl = git@github.com:Org/App.git\n", 0o644)
	write(filepath.Join(repo, ".claude", "settings.local.json"), `{"permissions":{"allow":["Bash(make)"]}}`, 0o644)
	write(filepath.Join(repo, "CLAUDE.local.md"), "# repo\n", 0o644)
	if _, err := core.RuleSet(home, repo, "work"); err != nil {
		t.Fatal(err)
	}
	return home, src
}

// El portal clasifica lo que captura un snapshot. Lo que no reconozca cae en
// «Otros», y eso no se ve: la pantalla sigue pintando, solo que el día que ccp
// capture un archivo nuevo aparecerá ahí sin nombre ni motivo. Por eso el
// inventario de rutas lo genera core y no una lista a mano.
func TestElPortalClasificaTodoLoQueSeCaptura(t *testing.T) {
	node, err := exec.LookPath("node")
	if err != nil {
		t.Skip("sin node: no se puede ejecutar el modelo del portal")
	}
	home, src := maquinaFalsa(t)
	srcs, err := core.SnapshotSources(home, src, core.SnapshotSourceOpts{})
	if err != nil {
		t.Fatal(err)
	}
	lpaths := make([]string, 0, len(srcs))
	for _, s := range srcs {
		lpaths = append(lpaths, s.LPath)
	}
	if len(lpaths) < 10 {
		t.Fatalf("la máquina falsa se quedó corta: %v", lpaths)
	}
	b, err := json.Marshal(map[string]any{"lpaths": lpaths})
	if err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(t.TempDir(), "config.json")
	if err := os.WriteFile(path, b, 0o600); err != nil {
		t.Fatal(err)
	}
	out, err := exec.Command(node, "config_test.mjs", path).CombinedOutput()
	t.Logf("%s", out)
	if err != nil {
		t.Fatalf("el portal no clasifica todo lo que se captura: %v", err)
	}
}

File: internal/cloud/portal/config_test.mjs
// Comprueba que el modelo de P-20 del portal reconoce todas las rutas lógicas
// que captura un snapshot. El inventario lo genera config_test.go desde core,
// no una lista a mano: así el día que ccp capture un archivo nuevo, esto falla
// en vez de enseñarlo como «Otros» sin explicación.
import { readFileSync } from 'node:fs';
import { classify, layersOf, itemsOf, byType, sections, applySection, TIPOS } from './web/js/config.js';

const v = JSON.parse(readFileSync(process.argv[2], 'utf8'));
let fallos = 0;
const check = (nombre, ok, extra) => {
  if (!ok) { fallos++; console.error('FALLA', nombre, extra ?? ''); } else { console.log('ok  ', nombre); }
};

const sinTipo = v.lpaths.filter((l) => classify(l).type === 'Otros');
check('ninguna ruta capturada cae en «Otros»', sinTipo.length === 0, sinTipo.join(', '));

const sinCapa = v.lpaths.filter((l) => classify(l).layer.kind === 'otros');
check('toda ruta capturada tiene capa', sinCapa.length === 0, sinCapa.join(', '));

const tiposRaros = v.lpaths.map((l) => classify(l).type).filter((t) => !TIPOS.includes(t));
check('los tipos están en la columna de P-20', tiposRaros.length === 0, tiposRaros.join(', '));

// Lo que no se puede editar tiene que decir por qué: un elemento gris sin
// motivo manda a buscar la explicación a otra pantalla.
const mudos = v.lpaths.map(classify).filter((c) => !c.editable && !c.reason);
check('lo no editable dice por qué', mudos.length === 0, mudos.map((c) => c.lpath).join(', '));

const items = v.lpaths.map((l) => ({ lpath: l, hash: 'h', size: 1, mode: 420, class: 'authored' }));
const capas = layersOf(items);
check('hay capa de ccp, global, perfil, proyecto y Desktop',
  ['ccp', 'global', 'perfil', 'proyecto', 'desktop'].every((k) => capas.some((c) => c.kind === k)),
  capas.map((c) => c.id).join(', '));
check('ccp y global van primero', capas[0].kind === 'ccp' && capas[1].kind === 'global', capas.map((c) => c.id).join(', '));

const perfil = capas.find((c) => c.kind === 'perfil' && c.name === 'work');
const suyos = itemsOf(items, perfil.id);
check('la capa de un perfil solo trae lo suyo', suyos.every((c) => c.lpath.startsWith('ccp/profiles/work/')), suyos.length);
check('los tipos salen agrupados y en orden', byType(suyos).every(([t]) => TIPOS.includes(t)));

// Los ajustes se abren por tipo; lo que no se reconoce viaja entero.
const texto = '{"model":"opus","permissions":{"allow":["Bash"]},"theme":"dark"}';
const secs = sections(texto);
check('un settings.json se abre en secciones',
  secs.map((s) => s.type).join(',') === 'Permisos,Modelo,Ajustes', JSON.stringify(secs));
check('un settings.json roto no se inventa secciones', sections('{ no') === null);
const tras = JSON.parse(applySection(texto, 'model', 'haiku'));
check('editar una sección no toca el resto',
  tras.model === 'haiku' && tras.theme === 'dark' && tras.permissions.allow[0] === 'Bash', JSON.stringify(tras));
const sinModelo = JSON.parse(applySection(texto, 'model', undefined));
check('borrar una sección la quita y deja el resto',
  sinModelo.model === undefined && sinModelo.theme === 'dark', JSON.stringify(sinModelo));

process.exit(fallos === 0 ? 0 : 1);
